src/Options.py

```diff
@@ -1,24 +1,117 @@
+from .hooks.Options import before_options_defined, after_options_defined, before_option_groups_created, after_option_groups_created
+from .Data import category_table, game_table, option_table
+from .Helpers import convertToLongString
 from .Locations import victory_names
 from .Items import item_table
 from .Game import starting_items
 
+from dataclasses import make_dataclass
+from typing import List
+import logging
+
 
 class FillerTrapPercent(Range):
     """How many fillers will be replaced with traps. 0 means no additional traps, 100 means all fillers are traps."""
     range_end = 100
 
-manual_options = before_options_defined({})
+def createChoiceOptions(values: dict, aliases: dict) -> dict:
+    values = {'option_' + i: v for i, v in values.items()}
+    aliases = {'alias_' + i: v for i, v in aliases.items()}
+    return {**values, **aliases}
 
+manual_options = before_options_defined({})
 manual_options["start_inventory_from_pool"] = StartInventoryPool
 
+manual_option_groups = {}
+manual_goal_override = {}
+
+# A list of Currently Supported Option types
+supported_option_types = ["Toggle", "Choice", "Range"]
+for option_name, option in option_table.get('data', {}).items():
+    if option_name.startswith('_'): #To allow commenting out options
+        continue
+
+    if option_name in ['goal', 'filler_traps']:
+        if manual_options.get('goal'):
+            logging.warning("Existing Goal option found created via Hooks, it will be overwritten by Manual's generated Goal option.\nIf you want to support old yaml you will need to add alias in after_options_defined")
+        if option_name == 'goal':
+            if option['type'] != 'Choice':
+                raise Exception("a 'goal' option must be of type 'Choice'")
+            manual_goal_override['args'] = createChoiceOptions({}, option.get('aliases', {}))
+            manual_goal_override['args']['default'] = args['default'] = option.get('default', 0)
+            manual_goal_override['args']['display_name'] = option.get('display_name', option_name)
+            manual_goal_override['description'] = convertToLongString(option.get('description', ''))
+        continue
+
+    if option_name not in manual_options:
+        option_type = option.get('type', "").title()
+
+        if option_type not in supported_option_types:
+            raise Exception(f'Option {option_name} in options.json has an invalid type of "{option["type"]}".\nIt must be one of the folowing: {supported_option_types}')
+
+        args = {'display_name': option.get('display_name', option_name)}
+
+        if option_type == "Toggle":
+            value = option.get('default', False)
+            option_class = DefaultOnToggle if value else Toggle
+
+        elif option_type == "Choice":
+            args = {**args, **createChoiceOptions(option.get('values'), option.get('aliases', {}))}
+            option_class = TextChoice if option.get("allow_user_input", False) else Choice
```

Comment:
I'll need to rename that because it sounds like it disable/enable the option or something
allow_custom_user_input might be a better parameter name